### PR TITLE
allow watching old resources with kubectl

### PR DESCRIPTION
**Pull request by sjenning**
Right now, one can not watch a resource with kubectl whose resourceVersion is outside the etcd watch window.  Specifying resourceVersion=0 returns the current object, then watches from the current index.

This PR changes the logic to use resourceVersion=0, which will work regardless of the resourceVersion of the object, and discard the first event if --watch-only is specified.

@ncdc @aveshagarwal

### CODE REVIEW

**Comment by aveshagarwal** [comment]
@kubernetes/rh-cluster-infra @kubernetes/kubectl

**Comment by smarterclayton** [comment]
This makes sense to me - it's a strict usability improvement (lots of people get broken by this when they try to watch infrequently updated objects).  (it's what resourceVersion=0 was added to enable)

**Comment by sjenning** [comment]
It wasn't passing unit tests with the resourceVersion=0 approach.  So I've completely changed it to do a GET the type in order to get the resourceVersion ~now (https://github.com/kubernetes/kubernetes/issues/16429#issuecomment-151976392). If a single pod is being watched, list the pods in order to get a PodList with the current resourceVersion, then watch on that resourceVersion that is ~now.

**Comment by smarterclayton** [comment]
That's not really a good fix - it's extremely inefficient, and I don't think we want to do that.  Can you discover why resourceVersion=0 doesn't pass tests?

**Comment by smarterclayton** [comment]
Literally `resourceVersion=0` was designed for this use case, so if you can't make it work we need to fix it.

**Comment by smarterclayton** [comment]
Note that there may be backwards compatibility issues (because older versions of the apiserver were broken with resourceVersion=0 when we introduced the watch cache), so some tests might just be broken because old code was broken.  Hard to say without knowing more.

**Comment by sjenning** [comment]
I'll take another look

**Comment by sjenning** [comment]
@smarterclayton this issue is  with the resourceVersion = 0 code returning the current state first, then watching at ~now:
https://github.com/kubernetes/kubernetes/blob/master/pkg/storage/watch_cache.go#L298

When watching a single resource, one can just drop the first event.  However, if a list of resources is being watched, there is an event per resource with no way, afaict, to distinguish between "initial state" events and "actual events that have occurred since the watch started".

**Comment by smarterclayton** [comment]
That comment is up for review.  I'm not sure we will make that change.

When you get the list of initial resources, the first resource version you see back is the "initial sync".  So object 1 has resource version X, then all subsequent objects that also have resource version X are in the initial sync and can be ignored.  So you always ignore the first item, and any subsequent item until you see a different resource version.

**Comment by sjenning** [comment]
@smarterclayton ok, new version posted. PTAL.  back to resourceVersion=0 approach and modifying the test cases to account for the "current state" events.  I also added a second pod to the test data and a new test for --watch-only on a resource list.

**Comment by ncdc** [comment]
@k8s-bot e2e test this issue: #IGNORE

File: pkg/kubectl/cmd/get.go [hunk 1]
```diff
@@ -188,12 +193,24 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
 		}
 
 		// print watched changes
-		w, err := r.Watch(rv)
+		w, err := r.Watch("0")
```
Comment (ncdc) [inline pkg/kubectl/cmd/get.go]: Maybe add a comment why you're using "0" and not rv here?

File: pkg/kubectl/cmd/get.go [hunk 2]
```diff
@@ -188,12 +193,24 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
...
 		}
 
 		kubectl.WatchLoop(w, func(e watch.Event) error {
+			accessor, err := meta.Accessor(e.Object)
+			if err != nil {
+				return fmt.Errorf("failed to create metadata accessor: %v", err)
+			}
+			eventrv, err := strconv.Atoi(accessor.GetResourceVersion())
+			if err != nil {
+				return fmt.Errorf("failed to convert resourceVersion string: %v", err)
```
Comment (ncdc) [inline pkg/kubectl/cmd/get.go]: Maybe display the value that failed to convert as well? `failed to convert resourceVersion string %q to an int: %v` or something like that?

**Comment by sjenning** [comment]
@smarterclayton @kubernetes/rh-ux anyone that can review, PTAL.  Thanks!

**Comment by janetkuo** [comment]
The unit test failed

```
FAIL	k8s.io/kubernetes/pkg/kubectl/cmd [build failed]
```

File: pkg/kubectl/cmd/get.go [hunk 3]
```diff
@@ -193,12 +198,26 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
 		}
 
 		// print watched changes
-		w, err := r.Watch(rv)
+		// using 0 instead of rv since rv might be outside the watch window
```
Comment (janetkuo) [inline pkg/kubectl/cmd/get.go]: minor: instead of rvstr

File: pkg/kubectl/cmd/get.go [hunk 4]
```diff
@@ -193,12 +198,26 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
...
+			accessor, err := meta.Accessor(e.Object)
+			if err != nil {
+				return fmt.Errorf("failed to create metadata accessor: %v", err)
+			}
+			eventrv, err := strconv.Atoi(accessor.GetResourceVersion())
+			if err != nil {
+				return fmt.Errorf("failed to convert resourceVersion string %q to an int: %v",
+					accessor.GetResourceVersion(), err)
+			}
+			if eventrv <= rv {
```
Comment (janetkuo) [inline pkg/kubectl/cmd/get.go]: We can't compare rv directly since it's not sequential.

File: pkg/kubectl/cmd/get.go [hunk 5]
```diff
@@ -180,7 +181,11 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
 		if err != nil {
 			return err
 		}
-		rv, err := mapping.MetadataAccessor.ResourceVersion(obj)
+		rvstr, err := mapping.MetadataAccessor.ResourceVersion(obj)
+		if err != nil {
+			return err
+		}
+		rv, err := strconv.Atoi(rvstr)
```
Comment (janetkuo) [inline pkg/kubectl/cmd/get.go]: We shouldn't assume that rv is a number.

**Comment by janetkuo** [comment]
cc @caesarxuchao

File: pkg/kubectl/cmd/get.go [hunk 6]
```diff
@@ -193,12 +198,26 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
...
 		if err != nil {
 			return err
 		}
 
 		kubectl.WatchLoop(w, func(e watch.Event) error {
+			accessor, err := meta.Accessor(e.Object)
+			if err != nil {
+				return fmt.Errorf("failed to create metadata accessor: %v", err)
+			}
+			eventrv, err := strconv.Atoi(accessor.GetResourceVersion())
```
Comment (liggitt) [inline pkg/kubectl/cmd/get.go]: Shouldn't need to convert to int. Clients should treat resourceVersion as opaque

Comment (caesarxuchao) [inline pkg/kubectl/cmd/get.go]: Agree. See https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L94-L98.

**Comment by sjenning** [comment]
@janetkuo @liggitt @ncdc I found another way to achieve the function not using resourceVersion as a number. PTAL when you can.

**Comment by sjenning** [comment]
@janetkuo @liggitt @smarterclayton can I get review on this?

File: pkg/kubectl/cmd/get.go [hunk 7]
```diff
@@ -182,9 +182,18 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
 		if err != nil {
 			return err
 		}
-		rv, err := mapping.MetadataAccessor.ResourceVersion(obj)
-		if err != nil {
-			return err
+
+		// watching from resourceVersion 0, starts the watch at ~now and
```
Comment (janetkuo) [inline pkg/kubectl/cmd/get.go]: Clarify what watching from rv 0 means and what problem it solves

File: pkg/kubectl/cmd/get_test.go [hunk 8]
```diff
@@ -693,6 +725,7 @@ func watchTestData() ([]api.Pod, []watch.Event) {
 				Spec: apitesting.DeepEqualSafePodSpec(),
 			},
 		},
+
```
Comment (janetkuo) [inline pkg/kubectl/cmd/get_test.go]: remove this line

File: pkg/kubectl/cmd/get_test.go [hunk 9]
```diff
@@ -740,10 +779,10 @@ func TestWatchSelector(t *testing.T) {
 	cmd.Flags().Set("selector", "a=b")
 	cmd.Run(cmd, []string{"pods"})
 
-	expected := []runtime.Object{&api.PodList{Items: pods}, events[0].Object, events[1].Object}
+	expected := []runtime.Object{podList, events[2].Object, events[3].Object}
 	actual := tf.Printer.(*testPrinter).Objects
 	if !reflect.DeepEqual(expected, actual) {
-		t.Errorf("unexpected object:\nExpected: %#v\n\nGot: %#v\n\n", expected[0], actual[0])
+		t.Errorf("unexpected object:\nExpected: %#v\n\nGot: %#v\n\n", expected, actual)
```
Comment (janetkuo) [inline pkg/kubectl/cmd/get_test.go]: This won't give anything useful for debugging: 
```
--- FAIL: TestWatchSelector (0.00s)
	get_test.go:787: unexpected object:
		Expected: []runtime.Object{(*api.PodList)(0xc8203e3320), (*api.Pod)(0xc8200f2780), (*api.Pod)(0xc8200f2a00)}
		
		Got: []runtime.Object{(*api.PodList)(0xc8203e3ec0), (*api.Pod)(0xc820390280), (*api.Pod)(0xc820390500)}
```

Comment (sjenning) [inline pkg/kubectl/cmd/get_test.go]: This actually just changes this case to match the others (see  ```TestWatchResource``` and ```TestWatchResourceIdentifiedByFile```). The information it conveys is basically a mismatch in the number lens of actual vs expected.  Dumping the actual structures not very useful either as the huge output is not diffable with tools.

**Comment by janetkuo** [comment]
Looks good now. Thanks!

**Pull request merged**